pkg/p2p/kadcli/reader.go

```diff
@@ -0,0 +1,120 @@
+// This Source Code Form is subject to the terms of the MIT License.
+// If a copy of the MIT License was not distributed with this
+// file, you can obtain one at https://opensource.org/licenses/MIT.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+package kadcli
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"errors"
+	"io"
+
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/checksum"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/protocol"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
+	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/eventbus"
+	"github.com/dusk-network/dusk-protobuf/autogen/go/rusk"
+)
+
+// Reader abstracts all of the logic and fields needed to receive messages from
+// other network nodes.
+type Reader struct {
+	processor *peer.MessageProcessor
+	gossip    *protocol.Gossip
+
+	cli  rusk.NetworkClient
+	stop context.CancelFunc
+}
+
+// NewReader makes a new kadcast reader that handles TCP packets of broadcasting.
+func NewReader(publisher eventbus.Publisher, g *protocol.Gossip, p *peer.MessageProcessor, rusk rusk.NetworkClient) *Reader {
+	return &Reader{
+		processor: p,
+		gossip:    g,
+		cli:       rusk,
+	}
+}
+
+// Listen starts accepting and processing stream data.
+func (r *Reader) Listen() {
+	// create stream handler
+	stream, err := r.cli.Listen(context.Background(), &rusk.Null{})
+	if err != nil {
+		log.Fatalf("open stream error %v", err)
+		return
+	}
+
+	// create a context
+	ctx, cancel := context.WithCancel(context.Background())
+	r.stop = cancel
+
+	// listen for messages
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				// receive a message
+				msg, err := stream.Recv()
+				if err == io.EOF {
+					// TODO: Notify someone? Retry mechanism?
+					return
+				} else if err != nil {
+					log.Fatalf("recv error %v", err)
+				}
+				// Message received
+				go r.processMessage(msg)
+			}
+		}
+	}()
+}
+
+// processMessage propagates the received kadcast message into the event bus.
+func (r *Reader) processMessage(message *rusk.Message) {
+	reader := bytes.NewReader(message.Message)
+
+	// read message (extract length and magic)
+	b, err := r.gossip.ReadMessage(reader)
+	if err != nil {
+		log.WithError(err).Warnln("error reading message")
```

Comment:
Anytime we report an error\warning  in `processMessage` we should add `WithField("r_addr", message.Metadata.SrcAddress)` so that we can trace back the malformed message.